Improve repcode handling in sequence extraction API

senhuang42 opened the pull request:
The repcode handling logic was slightly off in the original version of this API - this PR aims to fix it by: 1) Keeping track of a dedicated repcode array, and 2) Persist the repcodes array through each block. I've aimed to make the handling logic as clear to read as possible since we will need a similar process in the sequence compression API.

Test Plan:
- Manual spot checking (roundtrips silesia.tar with compression api)
- Modify the unit test. Previously, it turns out that setting the `litProbability` field to something like 0.5 in the `RDG_genBuffer()` function would almost always cause the unit test to fail. With this PR, it no longer does.

senhuang42 commented:
discovered some issues, fixing now

terrelln commented:
Can you use `repcodes_t` and `ZSTD_updateRep()` instead of re-implementing the logic? 

https://github.com/facebook/zstd/blob/7205e609a91258c9c4b4757be5dd55a1b2b8507f/lib/compress/zstd_compress_internal.h#L388-L411

senhuang42 commented:
> Can you use `repcodes_t` and `ZSTD_updateRep()` instead of re-implementing the logic?

Oh amazing, I guess it makes sense that the logic for this was centralized somewhere already

terrelln commented on lib/compress/zstd_compress.c:
```diff
@@ -2516,24 +2518,22 @@ static void ZSTD_copyBlockSequences(ZSTD_CCtx* zc)
         }
+        outSeqs[i].offset = rawOffset;
+        updatedRepcodes = ZSTD_updateRep(rep, seqStoreSeqs[i].offset - 1, seqStoreSeqs[i].litLength == 0);
+        ZSTD_memcpy(rep, updatedRepcodes.rep, sizeof(repcodes_t));
```
It would be slightly more efficient to do:
1. Copy `rep` into `updatedRepcodes` before the loop
2. Only used `updatedRepcodes` in the loop
3. Copy `updatedRepcodes` into `rep` after the loop

terrelln commented on lib/compress/zstd_compress.c:
```diff
@@ -2516,24 +2518,22 @@ static void ZSTD_copyBlockSequences(ZSTD_CCtx* zc)
-            outSeqs[i].offset = seqStoreSeqs[i].offset - ZSTD_REP_NUM;
         }
+        outSeqs[i].offset = rawOffset;
+        updatedRepcodes = ZSTD_updateRep(rep, seqStoreSeqs[i].offset - 1, seqStoreSeqs[i].litLength == 0);
```
Can you add a comment explaining why you need to use `offset - 1` here.

terrelln commented on lib/compress/zstd_compress.c:
```diff
@@ -2547,6 +2549,7 @@ static void ZSTD_copyBlockSequences(ZSTD_CCtx* zc)
     outSeqs[i].matchLength = outSeqs[i].offset = outSeqs[i].rep = 0;
     seqStoreSeqSize++;
 
+    ZSTD_memcpy(zc->blockState.nextCBlock->rep, updatedRepcodes.rep, sizeof(repcodes_t));
```
We shouldn't need this line. `nextCBlock->rep` is set by the block compressor already.

An assert that they are the same would be nice though.

Cyan4973 commented:
Seems there are still a few differences detected ?

senhuang42 commented:
> Seems there are still a few differences detected ?

Hmm, it seems like `ZSTD_updateRep()` is used only for the optimal parser and superblocks, and the other matchfinders have their own repcode finding logic. As such, I suspect at lower compression levels, the repcode table in the `blockState` might begin to diverge from the one that we are deriving in `copyBlockSequences()`.

Cyan4973 commented:
Indeed, faster strategies only care about the 1st (mostly) and 2nd (sometimes) repcode.
They probably to not track the 3rd one for speed consideration.

senhuang42 commented:
> Indeed, faster strategies only care about the 1st (mostly) and 2nd (sometimes) repcode.
> They probably to not track the 3rd one for speed consideration.

I've reverted the previous commit

terrelln reviewed (approved):
> I reverted the previous commit

That should be fine for correctness. However, I wonder if it might be better not to modify the repcodes. I wouldn't naively expect that `ZSTD_copyBlockSequences()` would modify the repcodes. If something changes in how zstd handles repcodes, e.g. the block compressor expects repcode 1 or 2 == 0 if they are out of bounds, this could mess that up.

senhuang42 commented:
> > I reverted the previous commit
> 
> That should be fine for correctness. However, I wonder if it might be better not to modify the repcodes. I wouldn't naively expect that `ZSTD_copyBlockSequences()` would modify the repcodes. If something changes in how zstd handles repcodes, e.g. the block compressor expects repcode 1 or 2 == 0 if they are out of bounds, this could mess that up.



> > I reverted the previous commit
> 
> That should be fine for correctness. However, I wonder if it might be better not to modify the repcodes. I wouldn't naively expect that `ZSTD_copyBlockSequences()` would modify the repcodes. If something changes in how zstd handles repcodes, e.g. the block compressor expects repcode 1 or 2 == 0 if they are out of bounds, this could mess that up.

Actually yeah, I like that approach - if we don't update it, the locally derived repcodes within the block might still diverge from the "truth", but at the end of each block, we have the same ending/starting point as the block compressor does.
